File: snippets/cpp/tree/trie.cpp
struct Trie {
    struct Node { int cnt; array<int, 26> to; };
    vector<Node> t; int sz;
    Trie() : t(1), sz(1) { t[0].to.fill(-1); }
    void insert(string &s) {
        int u = 0;
        for (char c : s) {
            int ch = c - 'a';
            if (t[u].to[ch] == -1) {
                t[u].to[ch] = sz++;
                t.pb({}); t.back().to.fill(-1);
            }
            u = t[u].to[ch];
        }
        t[u].cnt++;
    }
    bool search(string &s) {
        int u = 0;
        for (char c : s) {
            int ch = c - 'a';
            if (t[u].to[ch] == -1) return false;
            u = t[u].to[ch];
        }
        return t[u].cnt > 0;
    }
};

File: snippets/rust/tree/trie.rs
// Lowercase-letter trie. `cnt[u]` counts how many strings end at node `u`.
struct Trie {
    to: Vec<[i32; 26]>,
    cnt: Vec<i32>,
}
impl Trie {
    fn new() -> Self { Self { to: vec![[-1; 26]], cnt: vec![0] } }
    fn insert(&mut self, s: &[u8]) {
        let mut u = 0;
        for &c in s {
            let ch = (c - b'a') as usize;
            if self.to[u][ch] < 0 {
                self.to.push([-1; 26]);
                self.cnt.push(0);
                self.to[u][ch] = (self.to.len() - 1) as i32;
            }
            u = self.to[u][ch] as usize;
        }
        self.cnt[u] += 1;
    }
    fn search(&self, s: &[u8]) -> bool {
        let mut u = 0;
        for &c in s {
            let ch = (c - b'a') as usize;
            if self.to[u][ch] < 0 { return false; }
            u = self.to[u][ch] as usize;
        }
        self.cnt[u] > 0
    }
}
